File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    table = {}
    score = {}
    result = {}
        
    for friend in friends:
        table[friend] = {f:0 for f in friends}
        score[friend] = 0
        result[friend] = 0
        
    for gift in gifts:
        giver, receiver = gift.split(' ')
        
        table[giver][receiver] += 1
        
        score[giver] += 1
        score[receiver] -= 1
    
    for i in range(len(friends)-1):
        for j in range(i, len(friends)):
            
            more_gift = table[friends[i]][friends[j]] - table[friends[j]][friends[i]]
            
            if more_gift > 0:
                result[friends[i]] += 1
            elif more_gift < 0:
                result[friends[j]] += 1
            else:
                if score[friends[i]] > score[friends[j]]:
                    result[friends[i]] += 1
                elif score[friends[i]] < score[friends[j]]:
                    result[friends[j]] += 1
    
    answer = result[max(result, key=result.get)]
    return answer

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;

class Solution {
    public int solution(String[] friends, String[] gifts) {
        int answer = 0;
        Map<String, Integer> idxMap = new HashMap<>();
        int[] score = new int[friends.length];
        int[][] record = new int[friends.length][friends.length];
        
        for (int i = 0; i < friends.length; i++) {
            idxMap.put(friends[i], i);
        }
        
        for (String g : gifts) {
            String[] gift = g.split(" ");
            
            score[idxMap.get(gift[0])]++;
            score[idxMap.get(gift[1])]--;
            record[idxMap.get(gift[0])][idxMap.get(gift[1])]++;
        }

       
       for (int i = 0; i < friends.length; i++) {
           int cnt = 0;
           for (int j = 0; j < friends.length; j++) {
                if(i == j) continue;
               if (record[i][j] > record[j][i]) cnt++;
               else if (record[i][j] == record[j][i] && score[i] > score[j]) cnt++; 
           }
           answer = Math.max(cnt, answer);
       }
        return answer;
    }
}
